src/aspire/basis/fspca.py

```diff
@@ -282,6 +229,86 @@ def _compute_spca(self, coef):
         # the coefs.  This is used later for compression and index re-generation.
         self.sorted_indices = np.argsort(-np.abs(self.eigvals))
 
+        compressed_indices = self._get_compressed_indices()
+
+        self.spca_coef = np.zeros(
+            (self.src.n, len(compressed_indices)), dtype=self.dtype
+        )
+
+        # Compute coefficient vector of mean image at zeroth component
+        self.mean_coef_zero = self.mean_coef_est[self.angular_indices == 0]
+
+        # Define mask for zero angular mode, used in loop below
+        zero_ell_mask = self.basis._indices["ells"] == 0
+
+        # Apply Data matrix batchwise
+        num_batches = (self.src.n + self.batch_size - 1) // self.batch_size
+        for i in range(num_batches):
+
+            # Compute the coefficients for this batch
+            start = i * self.batch_size
+            finish = min((i + 1) * self.batch_size, self.src.n)
+            batch_coef = self.basis.evaluate_t(self.src.images[start:finish])
+
+            # Make the Data matrix (A_k)
+            # # Construct A_k, matrix of expansion coefficients a^i_k_q
+            # #   for image i, angular index k, radial index q,
+            # #   (around eq 31-33)
+            # #   Rows radial indices, columns image i.
+            # #
+            # # We can extract this directly (up to transpose) from
+            # #  fb coef matrix where ells == angular_index
+            # #  then use the transpose so image stack becomes columns.
+
+            # Initialize a totally empty BlkDiagMatrix, then build incrementally.
+            A = BlkDiagMatrix.empty(0, dtype=batch_coef.dtype)
+
+            # Zero angular index is special case of indexing.
+            A_0 = batch_coef[:, zero_ell_mask] - self.mean_coef_zero
+            A.append(A_0)
+
+            # Remaining angular indices have postive and negative entries in real representation.
+            for ell in range(
+                1, self.basis.ell_max + 1
+            ):  # `ell` in this code is `k` from paper
+                mask = self.basis._indices["ells"] == ell
+                mask_pos = [
```

Comment:
`mask_pos = mask & (self.basis._indices["sgns"] == +1)`?

Reply:
Yeah, I think that is better.  Changed.

Sorry, I lifted that directly from `covar2d.py` without too much thought. I'll make a separate pr for other one(s).

Thanks!